File: Aula 5/principal.py
from clima import Clima

def valor_temperatura(temp):
    temp = temp.strip().lower()  #remove espaços e coloca minusculo
    
    if temp == "frio":
        return 1
    elif temp == "ameno":
        return 2
    elif temp == "quente":
        return 3
    
def valor_preciptacao(prec):
    prec = prec.strip().lower() 

    if prec == "nada":
        return 1
    elif prec == "pouca":
        return 2
    elif prec == "média":
        return 3
    elif prec == "muita":
        return 4


lista = []  # cria a lista 
nome_base = "base.csv"  #importa a base de dados e define a variavel para armazena-la 

try: #tentar
    #abrindo o arquivo no modo leitura
    leitor = open(nome_base, "r", encoding="utf-8")

    for linha in leitor:
        ano, mes, temperatura, preciptacao = linha.strip().split(",") #fazendo o split, separando por ','

        obj_clima = Clima(ano, mes, temperatura, preciptacao) #criando o obj 

        if obj_clima not in lista:
            lista.append(obj_clima)
    
    leitor.close()
    
    for item in lista:
        print(item)

    dados_mes = {}

    for item in lista: 
        mes = item.mes

        dados_mes[mes] = {
            "soma_temp": 0, #armazena a soma das temperaturas 
            "soma_prec": 0, #armazena a soma das precipitações 
            "quantidade": 0 #armazena a quantidade de registros desse mes 
            }

        #soma os valores convertidos
        dados_mes[mes]["soma_temp"] += valor_temperatura(item.temperatura)
        dados_mes[mes]["soma_prec"] += valor_preciptacao(item.preciptacao)
        dados_mes[mes]["quantidade"] += 1   #armazena a quantidade de vezes que o mes apareceu 
    
    mes_mais_quente = None
    mes_mais_chuvoso = None 
    maior_media_temp = 0
    maior_media_prec = 0

    for mes in dados_mes: 
        media_temp = dados_mes[mes]["soma_temp"] / dados_mes[mes]["quantidade"]
        media_prec = dados_mes[mes]["soma_prec"] / dados_mes[mes]["quantidade"]     

        if media_temp > maior_media_temp:
            maior_media_temp = media_temp
            mes_mais_quente = mes 

        if media_prec > maior_media_prec:
            maior_media_prec = media_prec
            mes_mais_chuvoso = mes

    print("\n\tTemperatura:")
    print("Frio = 1 \nAmeno = 2 \nQuente = 3")
    print("\tPrecipitação:")
    print("Nada = 1\nPouca = 2\nMédia = 3\nMuita = 4")
    print(f"\nO mês que costuma fazer mais calor é {mes_mais_quente}, com média {maior_media_temp:.2f}.")
    print(f"O mês que costuma chover mais é {mes_mais_chuvoso}, com média {maior_media_prec:.2f}.")

except Exception as e:
    print("Ocorreu um erro...", e)  # o e serve pra mostra onde esta o erro 

File: Aula 5/Principal.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

public class Principal {

    public static int valorTemperatura(String temp) {
        if (temp.equals("Frio")) {
            return 1;
        } else if (temp.equals("Ameno")) {
            return 2;
        } else if (temp.equals("Quente")) {
            return 3;
        } else {
            return 0;
        }
    }

    public static int valorPrecipitacao(String prec) {
        if (prec.equals("nada")) {
            return 1;
        } else if (prec.equals("pouca")) {
            return 2;
        } else if (prec.equals("média")) {
            return 3;
        } else if (prec.equals("muita")) {
            return 4;
        } else {
            return 0;
        }
    }

    public static void main(String[] args) {
        List<Clima> lista = new ArrayList<>();
        String nomeBase = "base.csv";

        try (BufferedReader leitor = new BufferedReader(
                new FileReader(nomeBase, StandardCharsets.UTF_8))) {

            String linha;
            while ((linha = leitor.readLine()) != null) {
                String[] dadosLinha = linha.split(",");

                Clima objClima = new Clima(dadosLinha[0], dadosLinha[1], dadosLinha[2], dadosLinha[3]);

                if (!lista.contains(objClima)) {
                    lista.add(objClima);
                }
            }

            for (Clima item : lista) {
                System.out.println(item);
            }

            // HashMap para agrupar os dados por mês
            Map<String, int[]> dadosMes = new HashMap<>();

            // percorre a lista e agrupa por mês
            for (Clima item : lista) {
                String mes = item.mes;

                // se o mês ainda não existe no mapa, cria ele
                if (!dadosMes.containsKey(mes)) {
                    dadosMes.put(mes, new int[] { 0, 0, 0 });
                }

                //vetor para o valor de cada mes 
                int[] valores = dadosMes.get(mes);

                // [0] soma temperatura
                // [1] soma precipitação
                // [2] quantidade de registros
                // soma os valores convertidos
                valores[0] += valorTemperatura(item.temperatura);
                valores[1] += valorPrecipitacao(item.precipitacao);
                valores[2] += 1;
            }

            String mesMaisQuente = null;
            String mesMaisChuvoso = null;
            double maiorMediaTemperatura = 0;
            double maiorMediaPrecipitacao = 0;

            //cria o laço para percorrer todos os meses 
            for (String mes : dadosMes.keySet()) { //keySet serve para armazenar os meses 
                int[] valores = dadosMes.get(mes); //recupera o vetor com os dados acumulados do mes atual 

                double mediaTemp = (double) valores[0] / valores[2]; //faz a media da temperatura de cada mes 
                double mediaPrec = (double) valores[1] / valores[2];  //faz a media da precipitação de cada mes 

                System.out.println("\nMês: " + mes);
                System.out.println("Média da temperatura: " + mediaTemp);
                System.out.println("Média da precipitação: " + mediaPrec);

                if (mediaTemp > maiorMediaTemperatura) {
                    maiorMediaTemperatura = mediaTemp;
                    mesMaisQuente = mes;
                }

                if (mediaPrec > maiorMediaPrecipitacao) {
                    maiorMediaPrecipitacao = mediaPrec;
                    mesMaisChuvoso = mes;
                }
            }

            System.out.println("\nMês que costuma fazer mais calor: " + mesMaisQuente);
            System.out.println("Mês que costuma chover mais: " + mesMaisChuvoso);

        } catch (Exception e) {
            System.err.println("Ocorreu algum erro... " + e.getMessage());
        }
    }
}